ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+}
+
+void commit_allocation(heap_recorder *heap_recorder, heap_stack *heap_stack, VALUE obj, unsigned int weight) {
+  heap_record *heap_record = NULL;
+  if (!st_lookup(heap_recorder->heap_records, (st_data_t) heap_stack, (st_data_t*) &heap_record)) {
+    heap_record = heap_record_init(heap_stack);
+    if (st_insert(heap_recorder->heap_records, (st_data_t) heap_stack, (st_data_t) heap_record)) {
+      rb_raise(rb_eRuntimeError, "Duplicate heap stack tracking: %p", heap_stack);
+      return;
+    };
+  } else {
+    // FIXME: Figure out a way to not have to instantiate a new stack only to free it if it's
+    // already sampled. Something like supporting indexing the heap_records table with
+    // ddog_prof_Slice_Location objects directly for instance.
+    heap_stack_free(heap_stack);
+  }
+
+  object_record *object_record = object_record_init(obj, weight, heap_record);
+  if (st_insert(heap_recorder->object_records, (st_data_t) obj, (st_data_t) object_record) != 0) {
+    // Object already tracked?
+    object_record_free(object_record);
+    rb_raise(rb_eRuntimeError, "Duplicate heap object tracking: %lu", obj);
+    return;
+  }
+
+  fprintf(stderr, "Committed allocation of %lu (heap_record=%p, object_record=%p)\n", obj, heap_record, object_record);
+
+  heap_record->num_tracked_objects++;
+  heap_record->inuse_objects += weight;
+}
+
+void commit_free(heap_recorder *heap_recorder, VALUE obj) {
+  st_data_t key = (st_data_t) obj;
+  object_record *object_record = NULL;
+  if (!st_delete(heap_recorder->object_records, (st_data_t*) &key, (st_data_t*) &object_record)) {
+    // Object not tracked?
+    rb_raise(rb_eRuntimeError, "Committing free of untracked object");
+    return;
+  }
+
+  heap_record *heap_record = object_record->heap_record;
+  heap_record->num_tracked_objects--;
+  heap_record->inuse_objects -= object_record->weight;
+
+  fprintf(stderr, "Committed free of %lu (heap_record=%p, object_record=%p)\n", obj, heap_record, object_record);
+
+  object_record_free(object_record);
+}
+
+static void flush_queue(heap_recorder *heap_recorder) {
+  for (size_t i = 0; i < heap_recorder->queued_samples_len; i++) {
+    sample *queued_sample = &heap_recorder->queued_samples[i];
+    if (!queued_sample->skip) {
+      fprintf(stderr, "Flushing sample %p\n", queued_sample);
+
+      if (queued_sample->free) {
+        commit_free(heap_recorder, queued_sample->obj);
+      } else {
+        commit_allocation(heap_recorder, queued_sample->stack, queued_sample->obj, queued_sample->weight);
+      }
+    }
+
+    *queued_sample = EmptySample;
+  }
+  heap_recorder->queued_samples_len = 0;
+}
+
+static void enqueue_sample(heap_recorder *heap_recorder, sample new_sample) {
+  fprintf(stderr, "Enqueuing sample for %lu (weight=%u free=%i)\n", new_sample.obj, new_sample.weight, new_sample.free);
+  if (heap_recorder->queued_samples_len >= MAX_QUEUE_LIMIT) {
+    fprintf(stderr, "Dropping sample on the floor.\n");
+    return;
+  }
```

Comment:
This may be problematic if we're dropping a free. We could prioritize frees here (e.g. clobber allocations in the buffer), but we'd always hit a limit still.